rstudio/rstudio
# Add user option for selecting UI language

### Intent

Allow users to choose the language for the IDE UI.

Addresses #10455

### Approach

New user preference, "ui_language", appears as a dropdown in Global Options / General / Advanced, with "English" and "French" as options.

Changing this will cause the UI to reload to display the changed language.

![Screen Shot 2022-04-06 at 15 28 09](https://user-images.githubusercontent.com/10569626/162075465-372cda8b-968f-4a2d-8cee-73997c3d0fb7.png)

In order to make this value available to the web page when it is loading (before GWT has initialized), we also write the value to local storage in the web browser. You can see this via Chrome dev tools:

![Screen Shot 2022-04-06 at 15 29 15](https://user-images.githubusercontent.com/10569626/162075814-6f41866e-effb-4da7-ac9a-db3a8493fad3.png)

All modern browsers (including Electron and even our older QtWebEngine) support local storage, so I opted to use it instead of setting yet-another cookie.

GWT determines which language to show via a <meta> tag in the HTML page header, for example:

` <meta name="gwt:property" content="locale=fr">`

The index.htm page uses JavaScript to set this header, reading the value from local storage.

During session startup, if it is determined that the ui_language preference wasn't reflected in the local storage, then it will be set, and the page reloaded. This handles cases such as using a different browser, or if the admin has pre-configured the language to be non-English.

Also a small tweak to Electron code so it can find build output from CLion in dev configuration (folder starts with `cmake-build-` instead of `build`).

### Automated Tests

None

### QA Notes

Verify switching from English to French, in Server and Desktop (both Qt and Electron). Also try switching to French in one web browser, then loading the session from a different web browser that hasn't previously hit the server; should see the page start to load, then reload again, as it switches to French and writes the value to local storage. Subsequent loads should not require the reload.

You can also play with the value in local storage using browser dev tools, deleting it or modifying it, to ensure the IDE puts it back in sync with the user preference.

Finally, also try setting the default language to French for all users on a system. Do this by creating a `rstudio-prefs.json` in the system location for the given platform:

```json
{
    "ui_language": "fr"
}
```

On Linux and Mac, put in `/etc/rstudio/rstudio-prefs.json`. On Windows, it goes in `%PROGRAMDATA%\RStudio\rstudio-prefs.json`. More on this mechanism can be found here: https://docs.rstudio.com/ide/desktop-pro/settings.html#preferences

### Checklist
